board: emsdp: add a new arc based board support

IRISZZW opened the pull request:
add the support of emsdp board, a new arc based board.

The DesignWare® ARC® EM Software Development Platform (SDP) is a flexible platform
for rapid software development on ARC EM processor-based subsystems. It is intended
to accelerate software development and debug of ARC EM processors and subsystems for
a wide range of ultra-low power embedded applications such as IoT, sensor fusion,
and voice applications. 

The EMSDP includes an FPGA-based hardware board with commonly used peripherals and 
interfaces for extensibility. Downloadable platform packages containing different hardware configurations enable the board to be programmed with different ARC EM processors and 
subsystems. Now it has support for ARC EM4, EM5D, EM6, EM7D, EM9D and EM11D core 
configuration.

The details about the EMSDP board: [EMSDP](https://www.synopsys.com/dw/ipdir.php?ds=arc-em-software-development-platform)

This PR includes the following support:

- processor core support

- uart-driver

- gpio-driver

IRISZZW committed:
boards: arc: emsdp: add basic emsdp board support

* add basic emsdp board support

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

dbkinder commented on boards/arc/emsdp/board.dtsi:
```diff
@@ -0,0 +1,79 @@
+		};
+	};
+
+};
```
It's good to end files with a newline

IRISZZW replied:
already done

dbkinder commented on boards/arc/emsdp/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+it using openocd and gdb. The program can be download via the USB cable into
+the code and data memories.
+
+Release Notes
```
There's a heading here with no content?

IRISZZW replied:
removed

dbkinder commented on boards/arc/emsdp/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+Release Notes
+*************
+
+References
```
Same here (a heading with no content).   If you want the links below to actually show up in the output here, add a ``target-notes`` directive before the first link:
```
References
**********
.. target-notes::

.. _Digilent Pmod Modules: http://store.digilentinc.com/pmod-modules
.. _Putty website: http://www.putty.org
```

IRISZZW replied:
added

vonhust reviewed (changes requested):
pls squash some commits to have a better commit history

vonhust reviewed (approved):
Sanitycheck tests are ok.  It's ok for me

dbkinder reviewed (approved):
+1 for doc changes, thanks!

nashif reviewed (changes requested):
Please squash the last commit into where the changes have originated.

IRISZZW commented:
> Please squash the last commit into where the changes have originated.

already done  and rebased with the master branch

SebastianBoe commented on boards/arc/emsdp/Kconfig.board:
```diff
@@ -0,0 +1,11 @@
+	bool "EM Software Development Platform"
+	depends on SOC_ARC_EMSDP
+	help
+	  The ARC EM Software Development Platform (emsdp) is an FPGA based development platform intended to support ARC licenses in developing their software for the ARC EM processor family and ARC EM Subsystems. It has the support for ARC EM4, EM5D, EM6, EM7D, EM9D and EM11D processors. ARC EM Enhanced Security Package (ESP) and ARC EM Subsystems (DFSS, SCSS, DSS) are also supported.
```
The help text should have newlines, see other multi-line help texts for how to do indentation correctly.

IRISZZW replied:
already done.

SebastianBoe commented on soc/arc/snps_emsdp/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+zephyr_compile_options_ifdef(CONFIG_FLOAT -mfpu=fpus_all)
+elseif(CONFIG_SOC_EMSDP_EM11D)
+zephyr_compile_options(-mmpy-option=6)
+zephyr_compile_options_ifdef(CONFIG_FLOAT -mfpu=fpuda_all)
```
indentation

IRISZZW replied:
already done.

ruuddw commented on boards/arc/emsdp/emsdp_em4.yaml:
```diff
@@ -0,0 +1,12 @@
+identifier: emsdp
```
All the emsdp boards have the same identifier (emsdp), but these should be unique.
Please change into emsdp_em4 here, and _em6 etc. in the other board yaml files.

IRISZZW replied:
already done.

ruuddw reviewed (commented):
In principle ok, but one thing should still be fixed: unique identifiers for the board yaml files.

vonhust commented:
ping for merge

vonhust commented:
@ruudw, PR of EMSDP is ready,  pls review, approve and merge

andrewboie commented:
needs a rebase, otherwise this LGTM

IRISZZW commented:
> needs a rebase, otherwise this LGTM

done

andrewboie commented:
Error: Aborting due to non-whitelisted Kconfig warning 'warning: default on the choice symbol
CPU_ARCEM (defined at boards/arc/emsdp/Kconfig.defconfig:13, arch/arc/Kconfig:19) will have no
effect, as defaults do not affect choice symbols'. Note: If this warning doesn't point to an actual
problem, you can add it to the whitelist at the top of
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/kconfig/kconfig.py.

galak commented on boards/arc/emsdp/emsdp_em4.dts:
```diff
@@ -0,0 +1,44 @@
+	soc {
+
+		uart@f0004000 {
+			interrupts = <108 1>;
```
Is this really a board level change?  Feels like this should possible be in a soc dtsi file?

IRISZZW replied:
emsdp board is similar to em_starterkit board, it is FPGA based board which supports different core configurations, such as EM4,
EM5D, EM6, EM7D, EM9D, EM9D+ESP, EM11D. in different core configurations the interrupts number may be different. so we need to point it out in core configuration related dts file.

galak replied:
Since these are different core/SoC I think this should move into dts/arc/em{5d,6,7,9d,9d+esp,11d}.dtsi

vonhust replied:
Currently, most of differences of  different core are board-specific. Not easy to move into dts/arc/

We know there are lots of duplications,  plan to do a cleanup with other arc boards together in future.

galak commented on soc/arc/snps_emsdp/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_compile_options(-mcpu=${GCC_M_CPU} -mno-sdata)
+
+if(CONFIG_SOC_EMSDP_EM4)
```
Feels like this should get refactored between the other arc SoC's, but no issue.

IRISZZW replied:
different core configuration has different compile options.

galak commented on soc/arc/snps_emsdp/soc.c:
```diff
@@ -0,0 +1,12 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
if file is empty, why have it?

IRISZZW replied:
to make code organization structure uniformed with other arc core board.

galak replied:
we should remove the file.

IRISZZW replied:
removed

galak reviewed (dismissed):
Various comments and some changes.

IRISZZW committed:
board: emsdp: doc update and bug fixes

- update doc for different core configuration.

- fix some bugs in dts related files.

- add dts config and defconfig for different core configuration.

- end files with a newline in boards/arc/emsdp/board.dtsi

- remove unused head in boards/arc/emsdp/doc/index.rst

- ARC_MPU_VER in different core is fixed. so remove some useless code
  for ARC_MPU_VER judgements in Kconfig.defconfig.* files for emsdp

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

vonhust commented:
@galak @ruuddw @carlescufi  ping for merge if no other issues.

carlescufi commented:
@galak please merge at your convenience

vonhust commented:
ping for merge, @carlescufi @galak @ioannisg @nashif
